datafusion/physical-optimizer/src/progressive_evaluation/util.rs

```diff
@@ -0,0 +1,171 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use datafusion_common::{
+    error::Result,
+    tree_node::{Transformed, TreeNodeRecursion},
+};
+use datafusion_datasource::{
+    file_groups::FileGroup, file_scan_config::FileScanConfig, source::DataSourceExec,
+    PartitionedFile,
+};
+use datafusion_physical_expr::LexOrdering;
+use datafusion_physical_plan::{sorts::sort::SortExec, ExecutionPlan};
+
+use itertools::Itertools;
+
+use crate::progressive_evaluation::extract_ranges::extract_ranges_from_files;
+
+// This is an optimization for the most recent value query `ORDER BY time DESC/ASC LIMIT n`
+// See: https://github.com/influxdata/influxdb_iox/issues/12205
+// The observation is recent data is mostly in first file, so the plan should avoid reading the others unless necessary
+//
+/// This function is to split all files in the same ParquetExec into different groups/DF partitions and
+/// set the `preserve_partitioning` so they will be executed sequentially. The files will later be re-ordered
+/// (if non-overlapping) by lexical range.
+pub fn split_parquet_files(
+    plan: Arc<dyn ExecutionPlan>,
+    ordering_req: &LexOrdering,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
```

Comment:
Here is where we ungroup the file source, prior to any calculation of lexical ranges.

We do look at the lexical range, at the file src, to decide how to ungroup. But then it gets transformed in the DAG before it hits the SPM.